receiver/kafkareceiver/kafka_receiver_test.go

```diff
@@ -576,6 +580,77 @@ func TestNewMetricsReceiver(t *testing.T) {
 	})
 }
 
+func TestComponentStatus(t *testing.T) {
+	t.Parallel()
+	_, receiverConfig := mustNewFakeCluster(t, kfake.SeedTopics(1, "otlp_spans"))
+
+	statusEventCh := make(chan *componentstatus.Event, 10)
+	waitStatusEvent := func() *componentstatus.Event {
+		select {
+		case event := <-statusEventCh:
+			return event
+		case <-time.After(10 * time.Second):
+			t.Fatal("timed out waiting for status event")
+		}
+		panic("unreachable")
+	}
+	assertNoStatusEvent := func(t *testing.T) {
+		t.Helper()
+		select {
+		case event := <-statusEventCh:
+			t.Fatalf("unexpected status event received: %+v", event)
+		case <-time.After(100 * time.Millisecond):
+		}
+	}
+
+	// Create an intermediate TCP listener which will proxy the connection to the
+	// fake Kafka cluster. This can be used to verify the initial "OK" status is
+	// reported only after the broker connection is established.
+	lis, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+	t.Cleanup(func() { assert.NoError(t, lis.Close()) })
+	brokers := receiverConfig.Brokers
+	receiverConfig.Brokers = []string{lis.Addr().String()}
+
+	f := NewFactory()
+	r, err := f.CreateTraces(context.Background(), receivertest.NewNopSettings(metadata.Type), receiverConfig, &consumertest.TracesSink{})
+	require.NoError(t, err)
+	require.NoError(t, r.Start(context.Background(), &statusReporterHost{
+		report: func(event *componentstatus.Event) {
+			statusEventCh <- event
+		},
+	}))
+	t.Cleanup(func() {
+		assert.NoError(t, r.Shutdown(context.Background()))
+	})
+
+	// Connection to the Kafka cluster is asynchronous; the receiver
+	// will report that it is starting before the connection is established.
+	assert.Equal(t, componentstatus.StatusStarting, waitStatusEvent().Status())
+	// The StatusOK event should not be reported yet, as the connection to the
+	// fake Kafka cluster is not established yet.
+	assertNoStatusEvent(t)
+
+	// Accept the connection, proxy to the fake Kafka cluster.
+	conn, err := lis.Accept()
+	require.NoError(t, err)
+	defer conn.Close()
+	kfakeConn, err := net.Dial("tcp", brokers[0])
+	require.NoError(t, err)
+	defer kfakeConn.Close()
+	go func() { _, _ = io.Copy(conn, kfakeConn) }()
+	go func() { _, _ = io.Copy(kfakeConn, conn) }()
```

Comment:
Should this be tied to a waitgroup so we don't cause any potential flakes and ensure that these complete before the test is finished?

Reply:
Will do, though it's not strictly necessary: the connections will be closed by closing the client, which will terminate the copies. They won't be done by the time the function returns, but goleak will backoff/retry.